### PR TITLE
feat: Create Snackbar When Copying Schedules

### PR DESCRIPTION
**Pull request by KevinWu098**
## Summary
1. Open snackbar when copying (appending) schedules; happens on success and on failure

Snackbar opens on success

## Test Plan
1. Check that the snackbar opens appropriately on success and on failure

## Issues
Closes #

### CODE REVIEW

**Review by MinhxNguyen7 — APPROVED** [review]
LGTM!

File: apps/antalmanac/src/actions/AppStoreActions.ts [hunk 1]
```diff
@@ -182,13 +182,18 @@ export const changeCourseColor = (sectionCode: string, term: string, newColor: s
...
+export const copySchedule = (name: string, to: number) => {
     logAnalytics({
         category: analyticsEnum.addedClasses.title,
         action: analyticsEnum.addedClasses.actions.COPY_SCHEDULE,
     });
 
-    AppStore.copySchedule(to);
+    try {
+        AppStore.copySchedule(to);
+        openSnackbar('success', `Schedule copied to ${name}.`);
```
Comment (ap0nia) [inline apps/antalmanac/src/actions/AppStoreActions.ts]: Don't use this function.

**Review by ap0nia — CHANGES_REQUESTED** [review]

File: apps/antalmanac/src/actions/AppStoreActions.ts (hunk 1 above)
Comment (KevinWu098) [reply apps/antalmanac/src/actions/AppStoreActions.ts]: Hi, this looks like it's out of scope for the PR 🤓. It's scoped to just the snack bar, so let's move this to a separate PR 

🤓🤓🤓

Comment (ap0nia) [reply apps/antalmanac/src/actions/AppStoreActions.ts]: No, fix this now! 

Don't use the `openSnackbar` function because it's technically non-deterministic. The way it works is by extracting local component context into a global scope, which is bad because components have their own lifecycles and their local state doesn't inherently belong in global scope (this is different when using stores to share global state between components).

Prefer implementing functions with `onSuccess` and `onFailure` callbacks if you need to refer to the component's local stuff (in this case, `enqueueSnackbar` is a hook that grabs the snackbar context from the app, and it should remain local to components)

```tsx
export interface CallbackOptions {
  onSuccess: () => unknown
  onError: () => unknown
}

export const copySchedule = (name: string, to: number, options?: CallbackOptions) => {
    logAnalytics({
        category: analyticsEnum.addedClasses.title,
        action: analyticsEnum.addedClasses.actions.COPY_SCHEDULE,
    });

    // do logic

    if (success) {
      options?.onSuccess()
    } else {
      options?.onError()
    }
}
```

Any additional information needed by the component, i.e. the new schedules, messages, etc. can also be passed into the callback. It would be consumed as follows:

```tsx
function createCopyHandler(name, index, options) {
  return () => copySchedule(name, index,options)
}
function Component() {
  const { enqueueSnackbar } = useSnackbar()

  const options = useMemo(() => {
    return {
      onSuccess: () => enqueueSnackbar( ... )
    }
  }, [enqueueSnackbar])

  return (
    <MenuItem onClick={createCopyHandler(name, index, options} />
  )
}
```

Comment (ap0nia) [reply apps/antalmanac/src/actions/AppStoreActions.ts]: It's fairly verbose, but significantly more "idiomatic" because the responsibility of functions internal to components and external functions are kept separate. External functions just have to handle input and output, making them simpler to test and verify behavior. Component functions maintain their own separate logic to react to outputs generated by the external functions.

If it's too complicated, then I could take a look at it. I just kinda wanted to get contributors in a better mindset for using these tools in an "idiomatic" manner, rather than the scuffed way that's been pre-established.

Comment (KevinWu098) [reply apps/antalmanac/src/actions/AppStoreActions.ts]: Oh that's my bad -- I thought you were talking about `.copySchedule(to)` and got mixed up. I'll take a closer look at what you wrote up later today, but at first pass it makes sense

Will indeed fix it ~~now~~ _later_

**Comment by KevinWu098** [comment]
I'm not sure if my type definition for the CallbackOptions is proper, so just lmk if there's a better way to do them @ap0nia 

```ts
export interface CallbackOptions {
    onSuccess: (name?: string) => unknown;
    onError: (name?: string) => unknown;
}
```

File: apps/antalmanac/src/actions/AppStoreActions.ts [hunk 2]
```diff
@@ -12,6 +12,11 @@ import AppStore from '$stores/AppStore';
 import trpc from '$lib/api/trpc';
 import { courseNumAsDecimal } from '$lib/analytics';
 
+export interface CallbackOptions {
+    onSuccess: (name?: string) => unknown;
+    onError: (name?: string) => unknown;
```
Comment (ap0nia) [inline apps/antalmanac/src/actions/AppStoreActions.ts]: Could you just pass the numeric index? I don't think there's any need to modify the function signature in this way for this particular case.

```suggestion
    onSuccess: (index: number) => unknown;
    onError: (index: number) => unknown;
```

File: apps/antalmanac/src/actions/AppStoreActions.ts [hunk 3]
```diff
@@ -12,6 +12,11 @@ import AppStore from '$stores/AppStore';
 import trpc from '$lib/api/trpc';
 import { courseNumAsDecimal } from '$lib/analytics';
 
+export interface CallbackOptions {
```
Comment (ap0nia) [inline apps/antalmanac/src/actions/AppStoreActions.ts]: It depends on whether or not we want to reuse these callback options for other functions, i.e. all the helper functions that trigger the snackbar, but I think it's fine to specify that these options are specific to the `copySchedule` function.

```suggestion
export interface CopyScheduleOptions {
```

File: apps/antalmanac/src/actions/AppStoreActions.ts [hunk 4]
```diff
@@ -182,13 +187,18 @@ export const changeCourseColor = (sectionCode: string, term: string, newColor: s
...
         action: analyticsEnum.addedClasses.actions.COPY_SCHEDULE,
     });
 
-    AppStore.copySchedule(to);
+    try {
+        AppStore.copySchedule(to);
+        options?.onSuccess(name);
+    } catch (error) {
+        options?.onError(name);
+    }
```
Comment (ap0nia) [inline apps/antalmanac/src/actions/AppStoreActions.ts]:
```suggestion
    try {
        AppStore.copySchedule(to);
        options?.onSuccess(to);
    } catch (error) {
        options?.onError(to);
    }
```

File: apps/antalmanac/src/components/RightPane/AddedCourses/AddedCoursePane.tsx [hunk 5]
```diff
@@ -118,6 +119,14 @@ function ClearScheduleButton() {
 
 function CopyScheduleButton() {
     const [scheduleNames, setScheduleNames] = useState(AppStore.getScheduleNames());
+    const { enqueueSnackbar } = useSnackbar();
+
+    const options = useMemo(() => {
+        return {
+            onSuccess: (name?: string) => enqueueSnackbar(`Schedule copied to ${name}.`, { variant: 'success' }),
+            onError: (name?: string) => enqueueSnackbar(`Could not copy schedule to ${name}.`, { variant: 'error' }),
```
Comment (ap0nia) [inline apps/antalmanac/src/components/RightPane/AddedCourses/AddedCoursePane.tsx]:
```suggestion
            onSuccess: (index: number) => enqueueSnackbar(`Schedule copied to ${scheduleNames[index]}.`, { variant: 'success' }),
            onError: (index: number) => enqueueSnackbar(`Could not copy schedule to ${scheduleNames[index]}.`, { variant: 'error' }),
```

File: apps/antalmanac/src/components/RightPane/AddedCourses/AddedCoursePane.tsx [hunk 6]
```diff
@@ -148,12 +157,14 @@ function CopyScheduleButton() {
                             <MenuItem
                                 key={index}
                                 disabled={AppStore.getCurrentScheduleIndex() === index}
-                                onClick={createCopyHandler(index)}
+                                onClick={createCopyHandler(name, index, options)}
```
Comment (ap0nia) [inline apps/antalmanac/src/components/RightPane/AddedCourses/AddedCoursePane.tsx]:
```suggestion
                                onClick={createCopyHandler(index, options)}
```

**Review by ap0nia — CHANGES_REQUESTED** [review]
Thanks for implementing the strategy 🙏 

I think the original function signature should be completely preservable, i.e. it should accept just the index and possibly options. The options will be callback functions that will be called with the index after a successful/unsuccessful attempt to copy the schedule.

The callback functions can figure out the schedule name by indexing the array of schedule names with the index they were called with.

File: apps/antalmanac/src/actions/AppStoreActions.ts [hunk 7]
```diff
@@ -182,13 +187,18 @@ export const changeCourseColor = (sectionCode: string, term: string, newColor: s
     AppStore.changeCourseColor(sectionCode, term, newColor);
 };
 
-export const copySchedule = (to: number) => {
+export const copySchedule = (name: string, to: number, options?: CallbackOptions) => {
```
Comment (ap0nia) [inline apps/antalmanac/src/actions/AppStoreActions.ts]:
```suggestion
export const copySchedule = (to: number, options?: CallbackOptions) => {
```

**Comment by ap0nia** [comment]
I kinda wanna write tests for it, but I don't wanna write tests for it lol.

What do you wanna do Kevin? If you don't wanna test, then I'll just approve and merge 🤓

**Comment by KevinWu098** [comment]
> I kinda wanna write tests for it, but I don't wanna write tests for it lol.
> 
> What do you wanna do Kevin? If you don't wanna test, then I'll just approve and merge 🤓

If it breaks it breaks, and somebody gets extra contribution history 😋

**Comment by KevinWu098** [comment]
TLDR let's merge 😅, do you need me to do a noop?

**Review by ap0nia — APPROVED** [review]
Good work refactoring. This was more involved than it could've been otherwise, but I have to slow down Kevin from moving too fast 💯

**Comment by ap0nia** [comment]
Yeah I think I need an extra commit from someone aside from me lol. Unless you're able to merge it yourself.

**Review by ap0nia — APPROVED** [review]

**Comment by ap0nia** [comment]
image
lol it's not working

**Review by EricPedley — APPROVED** [review]
